### ISSUE TITLE
Remove public dependency to HDF5

### ISSUE THREAD

**Issue by rainman110** [dlr-gtlab/genh5#51]
In GitLab by @rainman110 on Feb 14, 2023, 21:30

There are still some includes to HDF 5, e.g. `H5public.h` in some headers. Are they required? 

If no, we would be able to use GenH5 completely without the HDF5 library from the users perspective, if we statically link HDF5 into GenH5.

**Comment by rainman110** [comment]
In GitLab by @mariusalexander on Mar 24, 2023, 14:59

@rainman110 

Most types can be wrapped rather easily:

 - `hid_t` which is simply a typedef for `long (long) int` depending on the platform.
 - Same goes for `hsize_t` which is similar to `size_t`.

 - Then there a many enum types which I probably can simply "copy" as well

Last but not least I am using some structs, which can be copied as well very easily:
 - `hvl_t` which is:
    ```cpp
    struct {
        size_t len;
        void * p; 
    };
    ```
 - `H5R_ref_t` which is: 
    ```cpp
    struct {
        union {
            uint8_t __data[H5R_REF_BUF_SIZE]; 
            // ^ globally defined as 64. I dont know if its equal on all platforms ^
            int64_t align;                   
        } u;
    };
    ```

I see the benefits, but now the question is... should I do that? Is it something that other libraries do as well? It just feels a bit wrong to simply copy all the definitions. 

I could maybe use the preprocessor to switch between these types, which must be defined to "deactive" the native hdf5 symbols?

**Comment by rainman110** [comment]
@mariusalexander This issue becomes now more important. 

We have conflicts with the hdf5 library provided by GTlab and the one used by python packages, e.g. `h5py` or `netCDF4`. This makes is really hard to use these packages in GTlab's python environment, as they typically use different DLL versions, leading to a crash of the python import.

And even with the same hdf5 version, it is currently very hard. Indeed the hdf5.dll from GTlab does not include all features required for h5py, so we basically need to copy h5py's hdf5.dll into the GTlab `bin` folder.

To fix this, we could statically link hdf5 into genh5, such that there is no hdf5.dll that can conflict with the python packages.

To to so however, all public uses of HDF5 need to be removed such that no part of GTlab or the HDF5 module may directly use HDF5. Instead, everything needs to be routed via GenH5.

@mariusalexander  Do you think, this will be possible?

And to answer your questions: Yes all hdf5 types need to be wrapper in Genh5 types in some way,

**Comment by mariusalexander** [comment]
Yes, I think I can do that. I will have to look at how the library defines certain symbols. What do you think about the idea to use a pre processor macro to be able to switch between dynamic and satic builds. If linked dynamically, the symbols of the HDF5 C-Library would be used, when linked statically, the "dummy/wrapped" HDF5 symbols are used
